Return 415 if invalid JSON

kathy-t opened the pull request:
**Description**
This PR returns a 415 if the search query is invalid JSON.

**Review Instructions**
Need to wait for the 1.14.1 hotfix to be deployed to staging before reviewing.

In your terminal, execute search commands with invalid JSON like the following:
- `curl -X POST https://staging.dockstore.org/api/api/ga4gh/v2/extended/tools/entry/_search -H 'accept: application/json' -d "foo"`
- `curl -X POST https://staging.dockstore.org/api/api/ga4gh/v2/extended/tools/entry/_search -H 'accept: application/json' -d "{"`

The response should be a 415 stating that the query is invalid JSON.

Also navigate to the search page in staging and sanity check that search is still functioning as expected (click on a few facets, search something).

**Issue**
[SEAB-3840](https://ucsc-cgl.atlassian.net/browse/SEAB-3840)

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead. 


[SEAB-3840]: https://ucsc-cgl.atlassian.net/browse/SEAB-3840?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kathy-t commented:
After discussion during stand-up, changed the target branch from develop to the 1.14.1 hotfix branch because it's a small change that may reduce community lead 500 investigations. Also changed the ticket's milestone to 1.14.1

coverbeck reviewed (commented):
I had an untested suggestion in #5493, "The endpoints should be declared to only consume application/json; I think if so declared, the call would fail in the framework before even getting to our code.".

It would be less code, plus the other advantage to that approach is that the request will show up as application/json in Swagger UI, and now that I think of it, the generated Swagger code, which might explain how one user perhaps innocently ran into this.

Untested though -- do you think it's worth a shot?

denis-yuen commented:
` @Produces`?
Worth a shot

coverbeck commented:
> ` @Produces`? Worth a shot

I think `@Consumes`

kathy-t commented:
> I think `@Consumes`

Thanks for the suggestion, I'll give it a try

kathy-t commented:
I added the `@Consumes` annotation, but turns out that's not enough to validate that the request body is valid JSON. After some testing and discussion with @hyunnaye, it appears that `@Consumes("application/json")` checks the `Content-Type` header of the request and if it's not `application/json`, it returns a 415. However, if the request does contain an `application/json` header, then the request is allowed through and the endpoint will still return a 500 if the JSON is not valid, so I kept the `try/catch` block except I changed the status code to 415.

Examples of requests that the `@Consumes` annotation rejects:
- `curl -X POST -H "Content-Type: text/plain" http://localhost:8080/api/ga4gh/v2/extended/tools/entry/_search -d "{"`
   - `Content-Type` is not `application/json`
- `curl -X POST http://localhost:8080/api/ga4gh/v2/extended/tools/entry/_search -d "{"`
   -  When there's no `Content-Type` header, it defaults to `application/x-www-form-urlencoded`

denis-yuen commented:
> After some testing and discussion with @hyunnaye, it appears that `@Consumes("application/json")` checks the `Content-Type` header of the request and if it's not `application/json`, it returns a 415.

Interesting, so its a bit of a trust system.

kathy-t commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsExtendedApi.java:
```diff
@@ -94,6 +94,7 @@ public Response toolsOrgGet(
+    @Consumes(MediaType.APPLICATION_JSON)
     @Produces({MediaType.APPLICATION_JSON})
     @ApiOperation(nickname = ToolsIndexSearch.OPERATION_ID, value = ToolsIndexSearch.SUMMARY, notes = ToolsIndexSearch.DESCRIPTION, response = String.class)
     @ApiResponses(value = {@ApiResponse(code = HttpStatus.SC_OK, message = ToolsIndexSearch.OK_RESPONSE, response = String.class)})
```
Interesting, I think I'll leave it as is because the message is more user friendly, but if others feel strongly about it then I can change it

svonworl commented:
> I added the `@Consumes` annotation, but turns out that's not enough to validate that the request body is valid JSON. After some testing and discussion with @hyunnaye, it appears that `@Consumes("application/json")` checks the `Content-Type` header of the request and if it's not `application/json`, it returns a 415. However, if the request does contain an `application/json` header, then the request is allowed through and the endpoint will still return a 500 if the JSON is not valid, so I kept the `try/catch` block except I changed the status code to 415.

IMHO, my reading of the HTTP status codes (https://www.rfc-editor.org/rfc/rfc9110.html#name-client-error-4xx) is that a 400 might be more appropriate if the `Content-Type` is `application/json` but the json is malformed.  But, it's definitely debatable, and either works, so no need to change.
